server/authz/rbac_test.go

```diff
@@ -0,0 +1,298 @@
+)
+
+const (
+	testPolicyMapPath          = "testData/test-policy-map.csv"
+	testPropertyFilePath       = "testData/test-rbac-conf.yaml"
+	testPropertyReloadFilePath = "testData/test-rbac-conf-reload.yaml"
+	testUrlPath                = "/test"
+	testEmail                  = "test@test.com"
+)
+
+var (
+	groupReadOnly  = []string{"role:readonly"}
+	groupAdmin     = []string{"role:admin"}
+	groupDefault   = []string{"role:dev"}
+	groupNamespace = []string{"role:namespace"}
+)
+
+// TestCreateAuthorizer is a test implementation of the NewCasbinObject function.
+func TestCreateAuthorizer(t *testing.T) {
+	authorizer, err := NewCasbinObject(WithPolicyMap(testPolicyMapPath), WithPropertyFile(testPropertyFilePath))
+	assert.NoError(t, err)
+
+	// Test that the authorizer is not nil
+	assert.NotNil(t, authorizer)
+
+	// Check the roles loaded from the policy map
+	roles := authorizer.enforcer.GetAllRoles()
+	assert.Equal(t, 2, len(roles))
+}
+
+// TestAuthorize is a test implementation of the Authorize functionality.
+// It tests that the user is authorized correctly for the given request.
+func TestAuthorize(t *testing.T) {
+	authorizer, err := NewCasbinObject(WithPolicyMap(testPolicyMapPath), WithPropertyFile(testPropertyFilePath))
+	assert.NoError(t, err)
+
+	// Test that the authorizer is not nil
+	assert.NotNil(t, authorizer)
+
+	// get a test gin context for GET request
+	getRequest := mockHttpRequest("GET", "")
+
+	// Test that the user is authorized for the GET request
+	readOnlyUser := mockUserInfo(testEmail, groupReadOnly)
+	isAuth := authorizer.Authorize(getRequest, readOnlyUser)
+	assert.True(t, isAuth)
+
+	// Test that the admin user is authorized for the GET request
+	adminUser := mockUserInfo(testEmail, groupAdmin)
+	isAuth = authorizer.Authorize(getRequest, adminUser)
+	assert.True(t, isAuth)
+
+	postRequest := mockHttpRequest("POST", "")
+
+	// Test that the ReadOnly user is not
+	// authorized for the POST request
+	isAuth = authorizer.Authorize(postRequest, readOnlyUser)
+	assert.False(t, isAuth)
+
+	// Test that the admin user is authorized for the POST request
+	isAuth = authorizer.Authorize(postRequest, adminUser)
+	assert.True(t, isAuth)
+}
+
+// TestDefaultPolicy is a test implementation of the DefaultPolicy functionality.
+// It tests that the default policy is set correctly when the authorizer is created.
+// Additionally, it tests that the default policy is applied correctly when a user is not found in the policy map.
+// The default policy is set to "role:readonly" in the test data.
+func TestDefaultPolicy(t *testing.T) {
+	authorizer, err := NewCasbinObject(WithPolicyMap(testPolicyMapPath), WithPropertyFile(testPropertyFilePath))
+	assert.NoError(t, err)
+
+	// Test that the authorizer is not nil
+	assert.NotNil(t, authorizer)
+
+	// Check the default policy, which is set to "role:readonly" in the test data
+	defaultPolicy := authorizer.getDefaultPolicy()
+	assert.Equal(t, "role:readonly", defaultPolicy)
+
```

Comment:
nit: I think we can consider removing this check as it's indirectly covered by the `Authorize` test right below. `getDefaultPolicy` is the implementation details of `Authorize` and we should aim to test what(behaviour) instead of how(implementation).

Reply:
Same for other test cases. We don't necessarily need to call `get` to verify policy or scope.

Reply:
Moved them to TestCreateAuthorizer